src/App/Application.cpp

```diff
@@ -649,118 +691,162 @@ std::vector<Document*> Application::openDocuments(const std::vector<std::string>
     for (auto &name : filenames)
         _pendingDocs.push_back(name.c_str());
 
-    std::map<Document *, DocTiming> newDocs;
+    std::map<DocumentT, DocTiming> timings;
 
     FC_TIME_INIT(t);
 
-    for (std::size_t count=0;; ++count) {
-        const char *name = _pendingDocs.front();
-        _pendingDocs.pop_front();
-        bool isMainDoc = count < filenames.size();
+    std::vector<DocumentT> openedDocs;
 
-        try {
-            _objCount = -1;
-            std::set<std::string> objNames;
-            if (_allowPartial) {
-                auto it = _pendingDocMap.find(name);
-                if (it != _pendingDocMap.end())
-                    objNames.swap(it->second);
-            }
+    int pass = 0;
+    do {

```

Comment:
Why use a `do...` here? Given the exit condition `!_pendingDocs.empty()`, you actively don't want this to guarantee the first iteration given your access to `_pendingDocs.front()`. In addition, the Core Guidelines advise against its use (see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es75-avoid-do-statements). I suggest switching to a `while...`:

Reply:
Well, I think this would be a 'genuine examples where a do-statement is a clear statement of a solution'. `_pendingDocs` may get pushed with new entry at the very end of this very long loop statement. And there is a big block of comments explaining the reason, about 10 lines above the exiting `while()`. Although this is not the only place where `_pendingDocs` is refilled, but I just want emphasize that push which is the solution to the problem that this PR is trying to solve.

Reply:
My concern here is that when someone reads this do...while loop, it calls their attention to the fact that you are trying to guarantee at least one execution, regardless of loop end condition. That is contrary to the code itself, however: if _pendingDocs were indeed empty upon entering this loop, you very much do NOT want to execute it once, the statement `_pendingDocs.front()` is almost sure to crash.

I argue that it's much more comfortable for a programmer reading this code to keep the loop end condition in mind as they read, despite the length of the loop. As written the only way to read it is to first scroll to the end, note the end condition, then scroll back to the top. Then you have to think "OK, why do we want to guarantee one execution of this loop, even if the end condition is satisfied?" And then you see line 704... it's very confusing. I don't think you accomplish your stated goal of emphasizing the final push: your large comment block emphasizes that without trouble: using this very non-idiomatic loop construction only muddies the waters.

Reply:
> My concern here is that when someone reads this do...while loop, it calls their attention to the fact that you are trying to guarantee at least one execution, regardless of loop end condition.

I think that's where our difference is. When I see a do {} while loop, I think of it as not worth it to check for loop entry condition because it is already guaranteed. And it is, as you can see with the few lines before the loop. Maybe it would be more obvious if I move the code to populate _pendingDocs closer together?

I see do {} while loop more for the emphasis on exit condition rather than entry, and the second half of this loop (which is newly added by this PR) is dedicated for checking the exit condition, so it would be logical to put the final check at the end.

Reply:
We can agree to disagree on this point: every developer is a little different, after all! I don't think this is a deal-breaker for the merge, it's not worth either of us spending more time considering it, IMO.